Tests/CacheAdvanceTests/CacheAdvanceTests.swift

```diff
@@ -97,34 +97,95 @@ final class CacheAdvanceTests: XCTestCase {
-            reader: try CacheReader(forReadingFrom: testFileLocation),
-            header: try CacheHeaderHandle(
-                forReadingFrom: testFileLocation,
-                maximumBytes: header.maximumBytes,
-                overwritesOldMessages: header.overwritesOldMessages),
-            decoder: JSONDecoder(),
-            encoder: JSONEncoder())
-
-        // Make sure the header data is persisted before we read it as part of the `messages()` call below.
-        try header.synchronizeHeaderData()
-        // Our file is empty. Make the file corrupted by setting the offset at end of newest message to be further in the file.
-        // This should never happen, but past versions of this repo could lead to a file having this kind of inconsistency if a crash occurred at the wrong time.
-        try header.updateOffsetInFileAtEndOfNewestMessage(
-            to: FileHeader.expectedEndOfHeaderInFile + 1)
-
+        let header = try prepareIncorrectOffsetInHeaderAndWriteMessage(messages: [],
+                                                                       randomHighValue: 101_000)
+        let cache = try CacheAdvance<TestableMessage>(fileURL: testFileLocation,
+                                                      maximumBytes: header.maximumBytes,
+                                                      shouldOverwriteOldMessages: header.overwritesOldMessages)
         XCTAssertThrowsError(try cache.messages()) {
             XCTAssertEqual($0 as? CacheAdvanceError, CacheAdvanceError.fileCorrupted)
         }
     }
 
+    func test_messages_timeout_WhenOffsetInFileAtEndOfNewestMessageLargerThanActualFileLength() throws {
+        do {
+            let message = TestableMessage(stringLiteral: TestableMessage.lorumIpsum[0].value)
+            let header = try prepareIncorrectOffsetInHeaderAndWriteMessage(messages: [message],
+                                                                           randomHighValue: 103_000)
+            let cacheToRead = try CacheAdvance<TestableMessage>(fileURL: testFileLocation,
+                                                                maximumBytes: header.maximumBytes,
+                                                                shouldOverwriteOldMessages: header.overwritesOldMessages)
+            DispatchQueue.global().async {
+                do {
+                    let _ = try cacheToRead.messages()
+                    // messages() should be in an infinite loop, and never return
+                    XCTFail()
+                } catch {
+                    XCTFail()
+                }
+            }
+        } catch {
+            XCTFail()
+        }
+        let expectation = XCTestExpectation()
+        DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
+            expectation.fulfill()
+        }
+        wait(for: [expectation], timeout: 2)
+    }
+
+  func test_messages_timeout_WhenOffsetInFileAtEndOfNewestMessageSmallThanActualFileLength() throws {
```

Comment:
@dfed  add another test with small offset value. It will cause the infinite loop too as we check like this.
```
        guard startingOffset != offsetInFileAtEndOfNewestMessage else {
            // We're at the last message.
            return nil
        }

```